usb: device_next: Remove experimental status

carlescufi opened the pull request:
The device_next stack now supports almost all classes that were covered by the legacy stack, and many controllers have been ported already.

See #42066 for more info.

carlescufi committed:
usb: device_next: Remove experimental status

The device_next stack now supports almost all classes that were covered
by the legacy stack, and many controllers have been ported already.

See #42066 for more info.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

tmon-nordic reviewed (commented):
Should we also come up with some API versions (e.g. 0.2.0) for USB components listed at https://docs.zephyrproject.org/latest/develop/api/overview.html?

Another thing would probably be removing the "Experimental DMA enabled" warning in DWC2 driver.

jfischer-no reviewed (changes requested):
I do not understand the reason for this. We do not support all classes. What exactly is the benefit of removing the EXPERIMENTAL? I see no reason to do this, at least not before v4.0.0 is tagged.

carlescufi commented:
> I do not understand the reason for this. We do not support all classes. What exactly is the benefit of removing the EXPERIMENTAL? I see no reason to do this, at least not before v4.0.0 is tagged.

The stack itself is more tested than many items that are today classed as stable. Support for many vendors has already been added. This is still not the default stack, but it certainly is beyond experimental from my point of view. My intention is to take this out of experimental before 4.0.0 but still it not be the default stack. After 4.0.0 and with all classes added, switch to this as the default.

carlescufi commented:
> Should we also come up with some API versions (e.g. 0.2.0) for USB components listed at https://docs.zephyrproject.org/latest/develop/api/overview.html?
> 
> Another thing would probably be removing the "Experimental DMA enabled" warning in DWC2 driver.

Yes, good idea. Let me add this to this PR tomorrow.

carlescufi committed:
usb: Clean up and set versions for the subsystem and drivers

Set the API versions and introduction release for the following
components:

- USBD subsystem
- USBH subsystem
- UDC drivers
- UHC drivers

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@tmon-nordic @jfischer-no I have now added versions to the APIs

kartben reviewed (changes requested):
this requires a more systemic "experimental USB" search and replace as this would leave quite a few documentation pages (mainly https://docs.zephyrproject.org/latest/connectivity/usb/index.html), sample READMEs, and Kconfigs behind.

carlescufi commented:
> this requires a more systemic "experimental USB" search and replace as this would leave quite a few documentation pages (mainly https://docs.zephyrproject.org/latest/connectivity/usb/index.html), sample READMEs, and Kconfigs behind.

Oops! Will do, thanks!

henrikbrixandersen commented:
> I do not understand the reason for this. We do not support all classes. What exactly is the benefit of removing the EXPERIMENTAL? I see no reason to do this, at least not before v4.0.0 is tagged.

I must admit I don't see the benefit or rationale behind this change either. From a user point of view, the new stack is not on par with the existing stack functionality-wise (i.e. USB device classes still needing to be ported). Taking it out of experimental could confuse users as to how far we are in making this an equal alternative to the existing device stack.

carlescufi commented:
> > I do not understand the reason for this. We do not support all classes. What exactly is the benefit of removing the EXPERIMENTAL? I see no reason to do this, at least not before v4.0.0 is tagged.
> 
> I must admit I don't see the benefit or rationale behind this change either. From a user point of view, the new stack is not on par with the existing stack functionality-wise (i.e. USB device classes still needing to be ported). Taking it out of experimental could confuse users as to how far we are in making this an equal alternative to the existing device stack.

The only class we are missing is DFU, which is nearing completion. Most of the drivers have been ported and the new stack has not only been thoroughly tested on Nordic's side, but also surpasses the old one in almost all metrics. My objective with this change is to give 4.0.0 users a fair and honest view of the state of the stack, which I think is currently misrepresented by the experimental tag.

jfischer-no commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -18,6 +18,15 @@
 #include <zephyr/sys/atomic.h>
 #include <zephyr/usb/usb_ch9.h>
 
+/**
+ * @brief USB device controller (UDC) API
+ * @defgroup udc_api USB device controller API
+ * @ingroup io_interfaces
+ * @since 3.3
+ * @version 0.2.0
+ * @{
+ */
```
I am not yet convinced to remove experimental for this release, but immediately after. But it is more work, there are many places where "experimental" needs to be removed and documentation needs to be updated.

There are few more headers to include API version, see https://github.com/zephyrproject-rtos/zephyr/pull/80385

jfischer-no commented on include/zephyr/usb/usbh.h:
```diff
@@ -26,9 +26,11 @@ extern "C" {
  * @defgroup usb_host_core_api USB Host Core API
  * @ingroup usb
+ * @since 3.3
+ * @version 0.2.0
```
This will remain experimental for the next release and should not be changed.

kartben commented:
@carlescufi will you come back to this? Would be great to get this out early enough in 4.1 IMO. Thanks!
